FIX: even admin does not see/like encrypted posts

lis2 opened the pull request:
Even admin should not see private and encrypted messaged.

Of course, admin can not decrypt messages so they are safe but right now they can like without seeing the content.

udan11 commented on lib/guardian_extensions.rb:
```diff
@@ -0,0 +1,15 @@
+
+module GuardianExtensions
+  def can_see_topic?(topic, hide_deleted = true)
+    if SiteSetting.encrypt_enabled? && topic.private_message? && topic.is_encrypted?
```
```suggestion
    if SiteSetting.encrypt_enabled? && topic&.is_encrypted?
```

I think you can simplify this as `is_encrypted?` already checks if `private_message?`.

Also, an admin could disable `discourse-encrypt` and bypass this check anyway. Is it for security purposes or did you want to short-circuit evaluation?

I believe `topic` can be `nil` too which might cause some problems. I see we handle this case in core:

```ruby
  def can_see_topic?(topic, hide_deleted = true)
    return false unless topic
    # ...
  end
```

eviltrout replied:
Rather than overwriting the method, I would recommend aliasing it. That way discourse-encrypt can call it, and, if true, check if encrypted as well. It's nicer than a freedom patch because if the underlying method changes you continue to get those updates.

udan11 commented on lib/guardian_extensions.rb:
```diff
@@ -0,0 +1,15 @@
+  end
+
+  def can_see_post?(post)
+    can_see_topic?(post.topic) && super
```
```suggestion
    can_see_topic?(post&.topic) && super
```

I believe `post` is allowed to be `nil` in this case too.

From core:

```ruby
  def can_see_post?(post)
    return false if post.blank?
    # ...
  end
```

udan11 commented on plugin.rb:
```diff
@@ -57,6 +58,7 @@ class DiscourseEncrypt::Engine < Rails::Engine
     User.class_eval              { prepend UserExtensions }
     Email::Sender.class_eval     { prepend EmailSenderExtensions }
     UserNotifications.class_eval { prepend UserNotificationsExtensions }
+    Guardian.class_eval      { prepend GuardianExtensions }
```
```suggestion
    Guardian.class_eval          { prepend GuardianExtensions }
```

Nit-pick: spacing here is a bit off.

udan11 reviewed (commented):
This looks good to me, great job! :+1:

I left a few comments, nothing major. I think we should handle the case when either `post` or `topic` are `nil`.

lis2 commented:
@udan11 could you take another look? I changed that a little bit. You are right that patching guardian is a scary thing so I introduced new guardian method and patched controllers to use it

udan11 reviewed (approved):
Looks good to me! :+1:
